1-theory/2-different-paradigms.adoc

```diff
@@ -0,0 +1,217 @@
+
+You need to layer metadata on top in order to describe things such as which fields correspond to which data types. This became part of the basis for SOAP, which used XML Schema and a WSDL (Web Services Description Language) to explain what went where and what it contained.
+
+This metadata is essentially what most science teachers drill into you from a young age: "label your units!" The sort of thing that stops people paying $100 for something that should have been $1 but was just marked as `price: 100` which was meant to be cents...
+
+All three of these specifications had implementations created by various people, mostly as open-source projects. Occasionally the folks who put the standard together created an official implementation in their favourite language, and the community built their own in other languages they wanted to use.
+
+A modern RPC specification is gRPC, which can easily be considered modern SOAP. It uses a data format called Protobuf, which requires a schema file as well as the data instance, much like the WSDL in SOAP. This Protobuf file is shared with both the client and the server, and then messages can be verified, and messages can be passed bewteen the server and client in binary, which leads to smaller messages than passing around big chunks JSON.
+
+gRPC focuses on making single RPC interactions, and it aims to achieve this as quickly as possible, thanks to the afore mentioned binary benefits, and its other huge benefit: HTTP/2. All of the gRPC implementations are HTTP/2 by default, and usually handle this with their own built in web server to make sure the HTTP/2 works the whole way through the transaction.
+
+== Representational State Transfer (REST)
+
+REST is a network paradigm described by Roy Fielding in a dissertation in 2000. REST is all about a client-server relationship, where server-side data are made available through representations of data in simple formats. This format is usually JSON or XML but could be anything (including Protobuf).
+
+These representations portray data from various sources as simple "resources", or "collections" of resources, which are then potentially modifiable with actions and relationships being made discoverable via a concept known as hypermedia controls (HATEOAS).
+
+Hypermedia controls are fundamental to REST, and is merely the concept of providing "next available actions", which could be related data, or more often it's actions available for that resource in its current state, like having a "pay" option for an invoice that has not yet been paid.
+
+These actions are just links, but the idea is the client knows that an invoice is payable by the presence of a "pay" link, and if that link is not there it should not show that option to the end user.
+
+{
+  "data"": {
+    "type": "invoice",
+    "id": "093b941d",
+    "attributes": {
+      "created_at": "2017–06–15 12:31:01Z",
+      "sent_at": "2017–06–15 12:34:29Z",
+      "paid_at": "2017–06–16 09:05:00Z",
```

Comment:
```suggestion

```

Reply:
Having the `paid_at` field and the `pay` link might be a bit confusing, since having this field in suggests the invoice has already been paid so why show the "pay" link? Or maybe this should be `"paid_at": "0000–00–00 00:00:00Z"`? Or maybe I am misunderstanding...

Reply:
IIIII fucked up.